mesonbuild/mtest.py

```diff
@@ -1851,36 +1855,51 @@ def test_suitable(self, test: TestSerialisation) -> bool:
 
         return True
 
-    def tests_from_args(self, tests: T.List[TestSerialisation]) -> T.Generator[TestSerialisation, None, None]:
+    @staticmethod
+    def make_selection_checker(patterns: T.List[str]) -> T.Callable[[TestSerialisation], bool]:
         '''
-        Allow specifying test names like "meson test foo1 foo2", where test('foo1', ...)
+        Allow specifying test names like "foo1", where test('foo1', ...)
 
         Also support specifying the subproject to run tests from like
-        "meson test subproj:" (all tests inside subproj) or "meson test subproj:foo1"
-        to run foo1 inside subproj. Coincidentally also "meson test :foo1" to
+        "subproj:" (all tests inside subproj) or "subproj:foo1"
+        to run foo1 inside subproj. Coincidentally also ":foo1" to
         run all tests with that name across all subprojects, which is
-        identical to "meson test foo1"
+        identical to "foo1"
         '''
-        for arg in self.options.args:
-            if ':' in arg:
-                subproj, name = arg.split(':', maxsplit=1)
+        unqualified = set()  # type: T.Set[str]
+        # None represents that any test in the subproject will match.
+        qualified = {}  # type: T.Dict[str, T.Union[T.Set[str], None]]
+
+        for pattern in patterns:
+            if ':' in pattern:
+                subproj, name = pattern.split(':', maxsplit=1)
             else:
-                subproj, name = '', arg
-            for t in tests:
-                if subproj and t.project_name != subproj:
-                    continue
-                if name and t.name != name:
-                    continue
-                yield t
+                subproj, name = '', pattern
+
+            if subproj == '':
+                unqualified.add(name)
+            elif name == '':
+                qualified[subproj] = None
```

Comment:
Might be bit nicer to use something like the following instead of `None` but I was not sure where to place it:

```
T = TypeVar('T')

class universe(Generic[T]):
    def __contains__(item: T) -> bool:
        return True
    def add(item: T) -> None:
        pass
```

Reply:
Added it to the current module for now.